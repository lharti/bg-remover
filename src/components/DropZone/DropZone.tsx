import { cn } from '@/utils/cn'
import { IconFileUpload } from '@tabler/icons-react'
import React from 'react'
import { useDropzone } from 'react-dropzone'

export interface DropZoneProps {
    onDrop: (acceptedFiles: File[]) => void
    className?: string
}

export const DropZone: React.FC<DropZoneProps> = ({ onDrop, className }) => {
    const { getRootProps, getInputProps, isDragActive } = useDropzone({
        accept: {
            'image/jpeg': [],
            'image/png': [],
            'image/bmp': [],
            'image/webp': [],
        },

        onDrop,
    })

    return (
        <div
            {...getRootProps()}
            aria-label="dropzone"
            className={cn(
                `
                  flex flex-col items-center justify-center rounded-lg border-2
                  border-dashed bg-white py-8 text-center text-gray-600
                  shadow-md transition-colors duration-300

                  hover:cursor-pointer hover:border-blue-400 hover:bg-blue-50
                  hover:text-blue-500

                  ${
                      isDragActive
                          ? 'border-blue-400 bg-blue-50'
                          : `border-gray-300`
                  }
                `,

                className,
            )}
        >
            <IconFileUpload
                size={50}
                className="mb-2 rounded-full bg-blue-100/40 p-2 text-blue-500"
            />

            <input {...getInputProps()} className="hidden" />

            {isDragActive ? (
                <p className="text-blue-800">{'Drop the images here...'}</p>
            ) : (
                <p className="font-medium">
                    {'Drag images here, or click to browse'}
                </p>
            )}
        </div>
    )
}
