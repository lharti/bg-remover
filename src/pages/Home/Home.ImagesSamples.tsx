import { SAMPLES } from '@/pages/Home/Home.constants'
import React from 'react'

interface ImagesSamplesProps {
    onSampleClick: (url: string) => Promise<void>
}
export const ImagesSamples: React.FC<ImagesSamplesProps> = ({
    onSampleClick: onSampleClick,
}) => {
    return (
        <div className="mt-10 space-y-4 text-center">
            <p className="font-bold">No images? Try one of these:</p>

            <div className="flex justify-center gap-4">
                {SAMPLES.map(({ name, thumbnail, url }) => (
                    <button
                        key={name}
                        className="hover:opacity-80"
                        role="button"
                        aria-label={`Select sample image: ${name}`}
                        onClick={() => onSampleClick(url)}
                    >
                        <img
                            src={thumbnail}
                            alt={name}
                            className={`w-40 rounded-lg object-cover shadow-lg`}
                        />
                    </button>
                ))}
            </div>
        </div>
    )
}
