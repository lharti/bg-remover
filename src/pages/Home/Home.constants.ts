export const ASSETS_CDN = 'https://assets.lharti.com/bg'

export const SAMPLES = [
    {
        name: 'bird',
        thumbnail: `${ASSETS_CDN}/upload-examples/bird-sm.jpg`,
        url: `${ASSETS_CDN}/upload-examples/bird-lg.jpg`,
    },

    {
        name: 'dog',
        thumbnail: `${ASSETS_CDN}/upload-examples/dog-sm.jpg`,
        url: `${ASSETS_CDN}/upload-examples/dog-lg.jpg`,
    },

    {
        name: 'fox',
        thumbnail: `${ASSETS_CDN}/upload-examples/fox-sm.jpg`,
        url: `${ASSETS_CDN}/upload-examples/fox-lg.jpg`,
    },
]
