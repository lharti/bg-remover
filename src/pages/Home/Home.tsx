import { DropZone } from '@/components/ui/DropZone'
import { generateImagesItemsFromFiles } from '@/libs/generateImageItemsFromFiles'
import { ImageItem } from '@/types/imageItem'
import React from 'react'
import { ImagesSamples } from './Home.ImagesSamples'

export interface HomePageProps {
    onImagesSelect: (files: ImageItem[]) => void
}

export const HomePage: React.FC<HomePageProps> = ({ onImagesSelect }) => {
    const onFilesUpload = (files: File[]) => {
        const imageItems = generateImagesItemsFromFiles(files)

        onImagesSelect(imageItems)
    }

    const onSampleClick = async (url: string) => {
        const response = await fetch(url)
        const blob = await response.blob()

        const file = new File([blob], 'sample.jpg', {
            type: 'image/jpeg',
        })

        onFilesUpload([file])
    }

    return (
        <div
            className={`
              mx-auto mt-16 flex max-w-2xl flex-col items-center justify-center
              space-y-12 px-4
            `}
        >
            <header className="text-center">
                <h1 className="text-4xl font-bold">AI Background Remover</h1>

                <p className="mt-4 text-lg">
                    Upload your images and remove backgrounds automatically
                </p>
            </header>

            <DropZone className="w-full py-10" onDrop={onFilesUpload} />

            <ImagesSamples onSampleClick={onSampleClick} />
        </div>
    )
}
