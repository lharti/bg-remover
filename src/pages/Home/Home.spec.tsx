import { DropZone } from '@/components/ui/DropZone'
import { HomePage } from '@/pages/Home'
import { render } from '@testing-library/react'

jest.mock('@/libs/generateImageItemsFromFiles')
jest.mock('@/components/ui/DropZone')

const DropZoneMock = jest
    .mocked(DropZone)
    .mockImplementation(({ className }) => (
        <div id="DropZoneMock" className={className} />
    ))

describe('<HomePage />', () => {
    it('should render', () => {
        expect.assertions(1)

        const onImagesSelect = jest.fn()

        const { container } = render(
            <HomePage onImagesSelect={onImagesSelect} />,
        )

        expect(container).toMatchSnapshot()
    })

    it('should pass uploaded images items to onImagesSelect onDrop', () => {
        expect.assertions(2)

        const onImagesSelect = jest.fn()

        render(<HomePage onImagesSelect={onImagesSelect} />)

        const onDrop = DropZoneMock.mock.calls[0][0].onDrop

        onDrop([
            new File([], 'cat.png', {
                type: 'image/png',
            }),
        ])

        expect(onImagesSelect).toHaveBeenCalledOnce()

        expect(onImagesSelect.mock.calls[0][0]).toMatchInlineSnapshot(`
            [
              {
                "id": "cat-UUID",
                "name": "cat",
                "original": "blob:http://test/cat.png",
                "status": "pending",
              },
            ]
        `)
    })
})
