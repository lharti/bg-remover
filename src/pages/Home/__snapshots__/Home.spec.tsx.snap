// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> should render 1`] = `
<div>
  <div
    class="
              mx-auto mt-16 flex max-w-2xl flex-col items-center justify-center
              space-y-12 px-4
            "
  >
    <header
      class="text-center"
    >
      <h1
        class="text-4xl font-bold"
      >
        AI Background Remover
      </h1>
      <p
        class="mt-4 text-lg"
      >
        Upload your images and remove backgrounds automatically
      </p>
    </header>
    <div
      class="w-full py-10"
      id="DropZoneMock"
    />
    <div
      class="mt-10 space-y-4 text-center"
    >
      <p
        class="font-bold"
      >
        No images? Try one of these:
      </p>
      <div
        class="flex justify-center gap-4"
      >
        <button
          aria-label="Select sample image: bird"
          class="hover:opacity-80"
          role="button"
        >
          <img
            alt="bird"
            class="w-40 rounded-lg object-cover shadow-lg"
            src="https://assets.lharti.com/bg/upload-examples/bird-sm.jpg"
          />
        </button>
        <button
          aria-label="Select sample image: dog"
          class="hover:opacity-80"
          role="button"
        >
          <img
            alt="dog"
            class="w-40 rounded-lg object-cover shadow-lg"
            src="https://assets.lharti.com/bg/upload-examples/dog-sm.jpg"
          />
        </button>
        <button
          aria-label="Select sample image: fox"
          class="hover:opacity-80"
          role="button"
        >
          <img
            alt="fox"
            class="w-40 rounded-lg object-cover shadow-lg"
            src="https://assets.lharti.com/bg/upload-examples/fox-sm.jpg"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
